platform/api/aws/ami.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aws
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+)
+
+type releaseAMIs struct {
+	AMIs []struct {
+		Name string `json:"name"`
+		PV   string `json:"pv"`
+		HVM  string `json:"hvm"`
+	}
+}
+
+func (a *API) AMI() (resolvedAMI string, resolveAMIErr error) {
+	a.opts.resolvedAMILock.Lock()
+	defer a.opts.resolvedAMILock.Unlock()
+
+	// Use cache if we can
+	if a.opts.ResolvedAMI != "" {
+		return a.opts.ResolvedAMI, nil
+	}
+
+	// If successfully resolved, cache the AMI id for next call.
+	// Decide based on the named returns
+	defer func() {
+		if resolveAMIErr == nil {
+			a.opts.ResolvedAMI = resolvedAMI
```

Comment:
The cache doesn't save any work unless an HTTP request is made, so for clarity this can be moved out of a defer to the point where the channel name is resolved to an AMI.